### PR TITLE
types: restore the field length of a string typed parameter value to the original one

### LINKED ISSUE #8644: unexpect data truncate error when prepare plan cache enabled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

enable prepare plan cache

```
mysql> create table t(data mediumblob);
Query OK, 0 rows affected (0.01 sec)

mysql> prepare stmt1 from 'insert t (data) values (?)';
Query OK, 0 rows affected (0.00 sec)

mysql> set @a = 'a';
Query OK, 0 rows affected (0.00 sec)

mysql> execute stmt1 using @a;
Query OK, 1 row affected (0.01 sec)

mysql> set @b = 'aaaaaaaaaaaaaaaaaa';
Query OK, 0 rows affected (0.00 sec)

mysql> execute stmt1 using @b;

```

2. What did you expect to see?

```
2 insert ok
```

3. What did you see instead?

```
ERROR 1406 (22001): Data too long for column 'data' at row 1
mysql> 
```

```
github.com/pingcap/errors.AddStack
	/home/robi/Code/go/pkg/mod/github.com/pingcap/errors@v0.11.0/errors.go:174
github.com/pingcap/parser/terror.(*Error).GenWithStack
	/home/robi/Code/go/pkg/mod/github.com/cloud-pi/parser@v0.0.0-20181211095502-1e4340922693/terror/terror.go:224
github.com/pingcap/tidb/types.ProduceStrWithSpecifiedTp
	/home/robi/Code/go/src/github.com/pingcap/tidb/types/datum.go:839
github.com/pingcap/tidb/types.(*Datum).convertToString
	/home/robi/Code/go/src/github.com/pingcap/tidb/types/datum.go:807
github.com/pingcap/tidb/types.(*Datum).ConvertTo
	/home/robi/Code/go/src/github.com/pingcap/tidb/types/datum.go:686
github.com/pingcap/tidb/expression.(*Constant).Eval
	/home/robi/Code/go/src/github.com/pingcap/tidb/expression/constant.go:102
github.com/pingcap/tidb/executor.(*InsertValues).evalRow
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert_common.go:247
github.com/pingcap/tidb/executor.(*InsertValues).insertRows
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert_common.go:189
github.com/pingcap/tidb/executor.(*InsertExec).Next
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert.go:142
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/adapter.go:288


[types:1406]Data too long for column 'data' at row 1
github.com/pingcap/errors.AddStack
	/home/robi/Code/go/pkg/mod/github.com/pingcap/errors@v0.11.0/errors.go:174
github.com/pingcap/parser/terror.(*Error).GenWithStack
	/home/robi/Code/go/pkg/mod/github.com/cloud-pi/parser@v0.0.0-20181211095502-1e4340922693/terror/terror.go:224
github.com/pingcap/tidb/executor.resetErrDataTooLong
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/write.go:171
github.com/pingcap/tidb/executor.(*InsertValues).handleErr
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert_common.go:215
github.com/pingcap/tidb/executor.(*InsertValues).evalRow
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert_common.go:248
github.com/pingcap/tidb/executor.(*InsertValues).insertRows
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert_common.go:189
github.com/pingcap/tidb/executor.(*InsertExec).Next
	/home/robi/Code/go/src/github.com/pingcap/tidb/executor/insert.go:142
github.com/pingcap/tidb/executor.(*ExecStmt).handleNoDelayExecutor
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

```
Release Version: None
Git Commit Hash: None
Git Branch: None
UTC Build Time: None
GoVersion: None
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false 
```

### PR DESCRIPTION
**Pull request by dbjoa**
### What problem does this PR solve? 
Fix #8644 

### What is changed and how it works?
* Restore the field length of a string typed parameter value to the original one

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - Need to cherry-pick to the release branch if the branch include the merged PR #8107

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8656)

### CODE REVIEW

**Comment by dbjoa** [comment]
@zhouqiang-cl Would you please check ``integration-ddl-test`` error?
```
ddl_test start

tidb_manager_ttl

10

/home/jenkins/workspace/tidb_ghpr_integration_ddl_test/go

go build github.com/pingcap/tidb/store: no non-test Go files in /home/jenkins/workspace/tidb_ghpr_integration_ddl_test/go/pkg/mod/github.com/pingcap/tidb@v0.0.0-20181207152923-6e2d6c7aa7eb/store

FAIL	github.com/pingcap/tidb-test/ddl_test/ddltest [build failed]

ddl_test end

script returned exit code 1
```

**Comment by zhouqiang-cl** [comment]
ok. I will check it

**Comment by dbjoa** [comment]
@zhouqiang-cl Thx!

File: types/field_type.go [hunk 1]
```diff
@@ -133,7 +133,11 @@ func DefaultParamTypeForValue(value interface{}, tp *FieldType) {
 		tp.Flen = UnspecifiedLength
 		tp.Decimal = UnspecifiedLength
 	default:
+		flen := tp.Flen
 		DefaultTypeForValue(value, tp)
+		if tp.Tp == mysql.TypeVarString {
```
Comment (zz-jason) [inline types/field_type.go]: Should we also check `mysql.TypeEnum`, `mysql.TypeSet`?

File: types/field_type.go [hunk 2]
```diff
@@ -133,7 +133,11 @@ func DefaultParamTypeForValue(value interface{}, tp *FieldType) {
...
+			tp.Flen = flen
```
Comment (eurekaka) [inline types/field_type.go]: Seems saved `flen` is always `UnspecifiedLength`, can we directly assign this to `tp.Flen` and remove line 136?

**Pull request merged**